pinot-common/src/main/java/com/linkedin/pinot/common/partition/PartitionAssignmentGenerator.java

```diff
@@ -0,0 +1,169 @@
+    for (int i = 0; i < numPartitions; i++) {
+      partitions.add(String.valueOf(i));
+    }
+
+    String tableNameWithType = tableConfig.getTableName();
+    int numReplicas = tableConfig.getValidationConfig().getReplicasPerPartitionNumber();
+    RealtimeTagConfig realtimeTagConfig = getRealtimeTagConfig(tableConfig);
+
+    List<String> consumingTaggedInstances = getConsumingTaggedInstances(realtimeTagConfig);
+    if (consumingTaggedInstances.size() < numReplicas) {
+      throw new IllegalStateException(
+          "Not enough consuming instances tagged. Must be atleast equal to numReplicas:" + numReplicas);
+    }
+    int maxConsumingServers = realtimeTagConfig.getMaxConsumingServers();
+    int numInstancesToUse = Math.min(maxConsumingServers, consumingTaggedInstances.size());
+
+    /**
+     * TODO: We will use only uniform assignment for now
+     * This will be refactored as AssignmentStrategy interface and implementations UniformAssignment, BalancedAssignment etc
+     * {@link PartitionAssignmentGenerator} and AssignmentStrategy interface will together replace
+     * StreamPartitionAssignmentGenerator and StreamPartitionAssignmentStrategy
+     */
+    return uniformAssignment(tableNameWithType, partitions, numReplicas, consumingTaggedInstances, numInstancesToUse);
+  }
+
+  /**
+   * Uniformly sprays the partitions and replicas across given list of instances
+   * Picks starting point based on table hash value. This ensures that we will always pick the same starting point,
+   * and return consistent assignment across calls
+   * @param allInstances
+   * @param partitions
+   * @param numReplicas
+   * @return
+   */
+  private PartitionAssignment uniformAssignment(String tableName, List<String> partitions, int numReplicas,
+      List<String> allInstances, int numInstancesToUse) {
+
+    PartitionAssignment partitionAssignment = new PartitionAssignment(tableName);
+
+    Collections.sort(allInstances);
+    List<String> instancesToUse = new ArrayList<>();
+    if (allInstances.size() <= numInstancesToUse) {
+      instancesToUse.addAll(allInstances);
+    } else {
+      int hashedServerId = Math.abs(EqualityUtils.hashCodeOf(tableName)) % allInstances.size();
+      for (int i = 0; i < numInstancesToUse; i++) {
+        instancesToUse.add(allInstances.get(hashedServerId++));
+        if (hashedServerId == allInstances.size()) {
+          hashedServerId = 0;
+        }
+      }
+    }
+
+    // pick a hashed start again, so that even in cases where we use all servers, we don't overload the servers at the start of the list
+    int hashedStartingServer = Math.abs(EqualityUtils.hashCodeOf(tableName)) % numInstancesToUse;
+    for (String partition : partitions) {
+      List<String> instances = new ArrayList<>(numReplicas);
+      for (int r = 0; r < numReplicas; r++) {
+        instances.add(instancesToUse.get(hashedStartingServer++));
+        if (hashedStartingServer == numInstancesToUse) {
```

Comment:
index  = (index+1) % numInstancesToUse  : makes it explicit assignment in one place